File: solutions/cpp/allergies/2/allergies.cpp
#include "allergies.h"

namespace allergies {
    allergy_test::allergy_test(int score_) {
        score = score_;
    };

    bool allergy_test::is_allergic_to(string ingredient) {
        std::unordered_set allergies = allergy_test::get_allergies();
        return allergies.find(ingredient) != allergies.end();
    }

    std::unordered_set<string> allergy_test::get_allergies() {
        vector<string> items = {"eggs", "peanuts", "shellfish", "strawberries", "tomatoes", "chocolate", "pollen", "cats"};

        int items_size = static_cast<int>(items.size());
        int subtotal = score % static_cast<int>(std::pow(2, items_size));
        std::unordered_set<string> allergies = {};

        // check for each ingredient in items list
        for (int i = 0; i < items_size; ++i) {
            if (subtotal >= std::pow(2, (items_size - 1 - i))) {
                allergies.insert(items[items_size - 1 - i]);
                subtotal -= std::pow(2, (items_size - 1 - i));
                }
            }
        return allergies;
        }

}  // namespace allergies

File: solutions/python/allergies/1/allergies.py
import itertools
import math

class Allergies:

    def __init__(self, score):
        self.score = score

    def allergic_to(self, item):
        return bool(item in self.lst)

    @property
    def lst(self):
        print(self.score)
        allergies_list = {
            1: 'eggs',
            2: 'peanuts',
            4: 'shellfish',
            8: 'strawberries',
            16: 'tomatoes',
            32: 'chocolate',
            64: 'pollen',
            128: 'cats'
        }

        # find summands of score
        allergies_shortlist = [k for k in allergies_list if k <= self.score]
        allergies_numbers = None
        done = False
        if self.score > 255: self.score = self.score - 2 ** int(math.log(self.score,2)) 
            
        for i in range(len(allergies_shortlist)+1):
            permutations = list(itertools.permutations(allergies_shortlist, i))
            for k in permutations:
                if sum(k) == self.score:
                    allergies_numbers = k
                    done = True
                    break
            if done == True: break

        output = [allergies_list[k] for k in allergies_numbers]
        return output